rootchain/samuel/samuel.go

```diff
@@ -0,0 +1,580 @@
+
+	if err != nil {
+		s.logger.Error(
+			fmt.Sprintf(
+				"Unable to extract event payload for SAM %s, %v",
+				SAM.Hash.String(),
+				err,
+			),
+		)
+	}
+
+	switch payloadType {
+	case rootchain.ValidatorSetPayloadType:
+		// Get the validator set info
+		vs, _ := rawPayload.(*payload.ValidatorSetPayload)
+		setInfo := vs.GetSetInfo()
+		validatorSetMap := make([]map[string][]byte, len(setInfo))
+
+		for index, info := range setInfo {
+			validatorSetMap[index] = map[string][]byte{
+				"Address":      info.Address,
+				"BLSPublicKey": info.BLSPublicKey,
+			}
+		}
+
+		// The method should have the signature
+		// methodName(validatorSet tuple[], index uint64, blockNumber uint64, signatures [][]byte)
+		encodedArgs, err := s.eventData.methodABI.Inputs.Encode(
+			map[string]interface{}{
+				"validatorSet":         validatorSetMap,
+				"index":                index,
+				"blockNumber":          blockNumber,
+				"signatures":           signatures,
+				"signatureBlockNumber": childBlockNumber,
+			},
+		)
+
+		if err != nil {
+			s.logger.Error(
+				fmt.Sprintf(
+					"Unable to encode method arguments for SAM %s, %v",
+					SAM.Hash.String(),
+					err,
+				),
+			)
+
+			return nil
+		}
+
+		return &types.Transaction{
+			Nonce:    0,
+			From:     types.ZeroAddress,
+			To:       &s.eventData.localAddress,
+			GasPrice: big.NewInt(0),
+			Gas:      framework.DefaultGasLimit,
+			Value:    big.NewInt(0),
+			V:        big.NewInt(1), // it is necessary to encode in rlp,
+			Input:    append(s.eventData.methodABI.ID(), encodedArgs...),
+		}
+	default:
+		s.logger.Error("Unknown payload type")
+	}
+
+	return nil
+}
+
+// PopReadyTransaction removes the latest ready transaction from the SAMP
+func (s *SAMUEL) PopReadyTransaction() {
+	s.samp.Pop()
+}
+
+// getVerifiedSAMBucket returns the verified SAM bucket that
+// has Quorum verified signatures
+func (s *SAMUEL) getVerifiedSAMBucket(
+	verifiedSAMs rootchain.VerifiedSAM,
+) rootchain.VerifiedSAM {
+	// Create the bucket map
+	// childchainBlockNum -> verifiedSAMs
+	samBuckets := make(map[uint64]rootchain.VerifiedSAM)
+
+	// Sort the SAM messages into buckets
+	for _, verifiedSAM := range verifiedSAMs {
+		childBlockNum := verifiedSAM.ChildBlockNum
+
+		// Check if there is already an aggregated array
+		samArr, present := samBuckets[childBlockNum]
+		if !present {
+			samArr = make(rootchain.VerifiedSAM, 0)
+		}
+
+		samArr = append(samArr, verifiedSAM)
+		samBuckets[childBlockNum] = samArr
+	}
+
+	var (
+		chosenBucket     uint64 = 0
+		candidateBuckets        = make([]uint64, 0)
+	)
+
+	// Get buckets that have Quorum signatures
+	for blockNum, sams := range samBuckets {
+		if uint64(len(sams)) >= s.signer.Quorum(blockNum) {
+			candidateBuckets = append(candidateBuckets, blockNum)
+		}
+	}
+
+	if len(candidateBuckets) == 0 {
+		// No candidate SAMs
+		return nil
+	}
+
+	// Out of all the candidate buckets, pick the one with the
+	// lowest quorum threshold
```

Comment:
Need more detailed reason here

Reply:
It actually doesn't matter which quorum bucket we choose - @dbrajovic and I just decided to go with the lowest possible quorum size to have the lowest transaction size